daemon/seccomp_linux_test.go

```diff
@@ -0,0 +1,198 @@
+		},
+		{
+			comment: "custom profile when seccomp is disabled returns error",
+			daemon: &Daemon{
+				seccompEnabled: false,
+			},
+			c: &container.Container{
+				SeccompProfile: "{ \"defaultAction\": \"SCMP_ACT_ERRNO\" }",
+				HostConfig: &config.HostConfig{
+					Privileged: false,
+				},
+			},
+			inSpec:  doci.DefaultLinuxSpec(),
+			outSpec: doci.DefaultLinuxSpec(),
+			err:     fmt.Errorf("seccomp is not enabled in your kernel, cannot run a custom seccomp profile"),
+		},
+		{
+			comment: "empty profile name loads default profile",
+			daemon: &Daemon{
+				seccompEnabled: true,
+			},
+			c: &container.Container{
+				SeccompProfile: "",
+				HostConfig: &config.HostConfig{
+					Privileged: false,
+				},
+			},
+			inSpec: doci.DefaultLinuxSpec(),
+			outSpec: func() coci.Spec {
+				s := doci.DefaultLinuxSpec()
+				profile, _ := seccomp.GetDefaultProfile(&s)
+				s.Linux.Seccomp = profile
+				return s
+			}(),
+		},
+		{
+			comment: "load container's profile",
+			daemon: &Daemon{
+				seccompEnabled: true,
+			},
+			c: &container.Container{
+				SeccompProfile: "{ \"defaultAction\": \"SCMP_ACT_ERRNO\" }",
+				HostConfig: &config.HostConfig{
+					Privileged: false,
+				},
+			},
+			inSpec: doci.DefaultLinuxSpec(),
+			outSpec: func() coci.Spec {
+				s := doci.DefaultLinuxSpec()
+				profile := &specs.LinuxSeccomp{
+					DefaultAction: specs.LinuxSeccompAction("SCMP_ACT_ERRNO"),
+				}
+				s.Linux.Seccomp = profile
+				return s
+			}(),
+		},
+		{
+			comment: "load daemon's profile",
+			daemon: &Daemon{
+				seccompEnabled: true,
+				seccompProfile: []byte("{ \"defaultAction\": \"SCMP_ACT_ERRNO\" }"),
+			},
+			c: &container.Container{
+				SeccompProfile: "",
+				HostConfig: &config.HostConfig{
+					Privileged: false,
+				},
+			},
+			inSpec: doci.DefaultLinuxSpec(),
+			outSpec: func() coci.Spec {
+				s := doci.DefaultLinuxSpec()
+				profile := &specs.LinuxSeccomp{
+					DefaultAction: specs.LinuxSeccompAction("SCMP_ACT_ERRNO"),
+				}
+				s.Linux.Seccomp = profile
+				return s
+			}(),
+		},
+		{
+			comment: "load prioritise container profile over daemon's",
+			daemon: &Daemon{
+				seccompEnabled: true,
+				seccompProfile: []byte("{ \"defaultAction\": \"SCMP_ACT_ERRNO\" }"),
+			},
+			c: &container.Container{
+				SeccompProfile: "{ \"defaultAction\": \"SCMP_ACT_LOG\" }",
+				HostConfig: &config.HostConfig{
+					Privileged: false,
+				},
+			},
+			inSpec: doci.DefaultLinuxSpec(),
+			outSpec: func() coci.Spec {
+				s := doci.DefaultLinuxSpec()
+				profile := &specs.LinuxSeccomp{
+					DefaultAction: specs.LinuxSeccompAction("SCMP_ACT_LOG"),
+				}
+				s.Linux.Seccomp = profile
+				return s
+			}(),
+		},
+	} {
+		t.Run(x.comment, func(t *testing.T) {
+			opts := WithSeccomp(x.daemon, x.c)
+			err := opts(nil, nil, nil, &x.inSpec)
+
+			assert.DeepEqual(t, x.inSpec, x.outSpec)
+			if !reflect.DeepEqual(err, x.err) {
+				t.Fatalf("expected:\n\t%v\n\ngot:\n\t%v", x.err, err)
+			}
```

Comment:
Ah, yes, I forgot the `nil` error case, which likely doesn't work with `assert.Error(...)`. I recall I opened a ticket for something better in gotest.tools 🤔  https://github.com/gotestyourself/gotest.tools/issues/153

Would `assert.DeepEqual()` work for that? 🤔 

```suggestion
			assert.DeepEqual(t, err, x.err)
```

Reply:
@thaJeztah sorry for taking this long to get back to you on this one. I got this error when I tried that:

![image](https://user-images.githubusercontent.com/5452977/105984952-bf8b4800-6092-11eb-8fb1-2c0701c7dc07.png)

Is it OK to keep the `reflect.DeepEqual(err, x.err)` validation?

Reply:
Ah, right, that's because the error returned probably uses `github.com/pkg/errors`, and has additional fields.

I think  for this case, it would be fine to keep things simple;

- change `err` in your `expected` to a `string`, containing the expected error-message
- if the expected error is not set (empty string); check that there's no error
- otherwise use `assert.Error(...)`;

```go
if x.err != "" {
	assert.Error(t, err, x.err)
} else {
	assert.NilError(t, err)
}
```

patch;

```patch
diff --git a/daemon/seccomp_linux_test.go b/daemon/seccomp_linux_test.go
index 22d26a7949..3b644986c1 100644
--- a/daemon/seccomp_linux_test.go
+++ b/daemon/seccomp_linux_test.go
@@ -3,8 +3,6 @@
 package daemon // import "github.com/docker/docker/daemon"

 import (
-       "fmt"
-       "reflect"
        "testing"

        coci "github.com/containerd/containerd/oci"
@@ -23,7 +21,7 @@ func TestWithSeccomp(t *testing.T) {
                c       *container.Container
                inSpec  coci.Spec
                outSpec coci.Spec
-               err     error
+               err     string
                comment string
        }

@@ -98,7 +96,7 @@ func TestWithSeccomp(t *testing.T) {
                        },
                        inSpec:  doci.DefaultLinuxSpec(),
                        outSpec: doci.DefaultLinuxSpec(),
-                       err:     fmt.Errorf("seccomp is not enabled in your kernel, cannot run a custom seccomp profile"),
+                       err:     "seccomp is not enabled in your kernel, cannot run a custom seccomp profile",
                },
                {
                        comment: "empty profile name loads default profile",
@@ -188,10 +186,11 @@ func TestWithSeccomp(t *testing.T) {
                t.Run(x.comment, func(t *testing.T) {
                        opts := WithSeccomp(x.daemon, x.c)
                        err := opts(nil, nil, nil, &x.inSpec)
-
                        assert.DeepEqual(t, x.inSpec, x.outSpec)
-                       if !reflect.DeepEqual(err, x.err) {
-                               t.Fatalf("expected:\n\t%v\n\ngot:\n\t%v", x.err, err)
+                       if x.err != "" {
+                               assert.Error(t, err, x.err)
+                       } else {
+                               assert.NilError(t, err)
                        }
                })
        }
```